lib/plausible/installation_support/verification/diagnostics.ex

```diff
@@ -0,0 +1,162 @@
+          plausible_is_initialized: true,
+          plausible_variant: plausible_variant,
+          test_event: %{
+            "requestUrl" => request_url,
+            "normalizedBody" => %{
+              "domain" => domain
+            },
+            "responseStatus" => response_status
+          },
+          service_error: nil
+        } = diagnostics,
+        data_domain,
+        url
+      )
+      when response_status in [200, 202] do
+    domain_is_expected? = domain == data_domain
+
+    tracker_is_version_2? = plausible_variant in @supported_variants
+    tracker_is_maybe_legacy? = is_nil(plausible_variant)
+
+    # in all envs, we want to succeed with verification of live websites
+    request_targets_plausible_api? =
+      String.starts_with?(request_url, "#{PlausibleWeb.Endpoint.url()}/api/event") or
+        String.starts_with?(request_url, "https://plausible.io/api/event")
+
+    request_targets_local_proxy? = String.starts_with?(request_url, "/")
+
+    cond do
+      (tracker_is_version_2? or tracker_is_maybe_legacy?) and
+        (request_targets_plausible_api? or request_targets_local_proxy?) and
+          domain_is_expected? ->
+        success()
+
+      (tracker_is_version_2? or tracker_is_maybe_legacy?) and
+        (request_targets_plausible_api? or request_targets_local_proxy?) and
+          not domain_is_expected? ->
+        error(@error_unexpected_domain)
+
+      true ->
+        interpret(diagnostics, data_domain, url)
+    end
+  end
+
+  @error_csp_disallowed Error.new!(%{
+                          message: "We encountered an issue with your site's CSP",
+                          recommendation:
+                            "Please add plausible.io domain specifically to the allowed list of domains in your Content Security Policy (CSP)",
+                          url:
+                            "https://plausible.io/docs/troubleshoot-integration#does-your-site-use-a-content-security-policy-csp"
+                        })
+  def interpret(
+        %__MODULE__{
+          plausible_is_on_window: true,
+          plausible_is_initialized: true,
+          test_event: %{
+            "requestUrl" => _request_url,
+            "error" => %{message: _error_message}
+          },
+          disallowed_by_csp: true,
+          service_error: nil
+        },
+        _data_domain,
+        _url
+      ) do
+    error(@error_csp_disallowed)
+  end
+
+  @unknown_error Error.new!(%{
+                   message: "Your Plausible integration is not working",
+                   recommendation:
+                     "Please manually check your integration to make sure that the Plausible snippet has been inserted correctly",
+                   url:
+                     "https://plausible.io/docs/troubleshoot-integration#how-to-manually-check-your-integration"
+                 })
+  def interpret(%__MODULE__{service_error: _service_error} = diagnostics, _data_domain, url) do
```

Comment:
Looks like matching on `%__MODULE__{service_error: _service_error}` doesn't have any effect here (the struct is always guaranteed to have the `service_error` key).

```suggestion
  def interpret(%__MODULE__{} = diagnostics, _data_domain, url) do
```